### PR TITLE
Scripts/Ulduar: Boss Kologarn

### LINKED ISSUE #20889: Kologarn arms issue
**Description:**
When the wipe happens to Kologarn, the arms should respawn only if dead.
CHANGEME Description of the problem or issue here.

**Current behaviour:**
Arms spawn no matter what when you die to Kologarn. This means if you wipe 5 times, you'll get 10 arms.

**Expected behaviour:**
When the wipe happens to Kologarn, the arms should respawn only if dead.
CHANGEME Description of the problem or issue here.
**Steps to reproduce the problem:**

1. Die to Kologarn.
2. Respawn
3. Die again
4. Tons of arms

**Branch(es):**

3.3.5a

TrinityCore rev. 7fd8603ec5e6+ 2017-11-16 18:35:58 +0100

**TDB version:**  CHANGEME Version of the TrinityCore database
TDB 61
**Operating system:** 
Windows 10

### PR DESCRIPTION
**Pull request by sirikfoll**
**Changes proposed:**

-  Script and enable all boss achievements;
-  Fix issues with Arms and evade;
-  Fix all visual issues with Focused Eyebeam spell;
-  Added missing spells;
-  Script Npc Rumble;
-  Add correct boss spawn, by areatrigger;
- Implemented Kologarn Pit Kill bunny;
-  Updated script register model.

**Target branch(es):** 3.3.5/master
- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #20889 #19445

**Tests performed:** Built and tested with 1 player

### CODE REVIEW

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+DELETE FROM `creature_template_movement` WHERE `CreatureId`=33661;
+INSERT INTO `creature_template_movement` (`CreatureId`, `Ground`, `Swim`, `Flight`, `Rooted`) VALUES
+(33661,0,0,0,1);
+
+DELETE FROM `creature_template_addon` WHERE `entry` IN(32930,33909,32934,32933);
+INSERT INTO `creature_template_addon` (`entry`,`path_id`,`mount`,`bytes1`,`bytes2`,`emote`,`auras`) VALUES
+(32934,0,0,0,1,0,64651),
+(32933,0,0,0,1,0,64651),
+(32930,0,0,0,1,0,64651),
+(33909,0,0,0,1,0,64651);
```
Comment (Killyana) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: "64651"

Comment (illfated) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Oh. Right. `auras` are `text` type, not int.

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp [hunk 2]
```diff
@@ -17,350 +17,334 @@
...
-                    default:
-                        return;
-                }
+        instance->DoStartTimedAchievement(ACHIEVEMENT_TIMED_TYPE_EVENT, CRITERIA_DISARMED);
+    }
 
-                summon->CastSpell(summon, SPELL_FOCUSED_EYEBEAM_PERIODIC, true);
-                summon->CastSpell(summon, SPELL_FOCUSED_EYEBEAM_VISUAL, true);
-                summon->SetReactState(REACT_PASSIVE);
+    void SetData(uint32 type, uint32 data) override
```
Comment (illfated) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp]: `unused parameter 'data' [-Wunused-parameter]`

**Comment by PolarCookie** [comment]
Tested with 2 characters, certainly an improvement
Found the following
1) Rubble sometimes spawn below the floor.
2) Eye beam might target the tank, should always pick a ranged if any before melee.
3) Stone Grip does not release the target when it dies.

**Comment by Shauren** [comment]
5. no

**Comment by Keader** [comment]
@PolarCookie in future, if you got some bug related with infinity combat, use command .debug combat, it will show what is stucked with you

**Comment by PolarCookie** [comment]
.debug combat gave
```
Combat refs: (Combat state: 1 |Manager state: 1)
[PvE] Armsweep Stalker Kologarn (SpawnID 136520)
```

**Comment by Cyclone47** [comment]
Great work so far!

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp [hunk 3]
```diff
@@ -373,304 +356,352 @@ class StoneGripTargetSelector : public std::unary_function<Unit*, bool>
...
+    }
+
+    void FilterTargetsInitial(std::list<WorldObject*>& unitList)
+    {
+        Creature* caster = GetCaster()->ToCreature();
+        // Remove "main tank" and non-player targets
+        unitList.remove_if(StoneGripTargetSelector(caster, caster->GetThreatManager().GetCurrentVictim()));
+        // Maximum affected targets per difficulty mode
+        uint32 maxTargets = 1;
+        if (GetSpellInfo()->Id == 63981)
```
Comment (Krudor) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp]: Make enum value for 63981

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp [hunk 4]
```diff
@@ -373,304 +356,352 @@ class StoneGripTargetSelector : public std::unary_function<Unit*, bool>
...
+            maxTargets = 3;
+
+        Trinity::Containers::RandomResize(unitList, maxTargets);
+        if (!unitList.empty())
         {
-            PrepareSpellScript(spell_ulduar_stone_grip_cast_target_SpellScript);
+            caster->AI()->Talk(SAY_GRAB_PLAYER);
+            caster->AI()->Talk(EMOTE_STONE_GRIP);
```
Comment (Krudor) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp]: Can you move this outside the target filter?

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp [hunk 5]
```diff
@@ -373,304 +356,352 @@ class StoneGripTargetSelector : public std::unary_function<Unit*, bool>
...
+    //! What we do here is remove all harmful aura's related and teleport to safe spot.
+    void OnRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        if (GetTargetApplication()->GetRemoveMode() != AURA_REMOVE_BY_ENEMY_SPELL)
+            return;
+
+        if (!GetOwner()->ToCreature())
+            return;
+
+        uint32 rubbleStalkerEntry = (GetOwner()->GetMap()->GetDifficulty() == DUNGEON_DIFFICULTY_NORMAL ? 33809 : 33942);
```
Comment (Krudor) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_kologarn.cpp]: Enum values for 33809 and 33942

**Comment by Keader** [comment]
@sirikfoll This branch has conflicts that must be resolved

**Comment by jackpoz** [comment]
any plan about resuming this PR ?

**Comment by jackpoz** [comment]
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.

**Pull request closed without merging**